ngmix/tests/test_metacal_bootstrap.py

```diff
@@ -0,0 +1,155 @@
+    return pr
+
+
+@pytest.mark.parametrize('noise', [1.0e-8, 0.01])
+@pytest.mark.parametrize('use_bootstrapper', [False, True])
+@pytest.mark.parametrize('nband', [None, 2])
+@pytest.mark.parametrize('nepoch', [None, 2])
+def test_metacal_bootstrap_max_smoke(noise, use_bootstrapper, nband, nepoch):
+    """
+    Smoke test a Runner running the LM fitter
+    """
+
+    rng = np.random.RandomState(2830)
+
+    data = get_model_obs(
+        rng=rng,
+        model='gauss',
+        noise=noise,
+        nepoch=nepoch,
+        nband=nband,
+    )
+    obs = data['obs']
+
+    prior = get_prior(
+        rng=rng,
+        cen=[0.0, 0.0],
+        cen_width=1.0,
+        T_range=[-1.0, 1.e3],
+        F_range=[0.01, 1000.0],
+        nband=nband,
+    )
+
+    flux_guess = data['gmix'].get_flux()
+    Tguess = data['gmix'].get_T()
+    guesser = TFluxAndPriorGuesser(
+        rng=rng, T=Tguess, flux=flux_guess, prior=prior,
+    )
+    psf_guesser = SimplePSFGuesser(rng=rng)
+
+    fitter = LM(model="gauss", prior=prior)
+    psf_fitter = LM(model='gauss')
+
+    psf_runner = PSFRunner(
+        fitter=psf_fitter,
+        guesser=psf_guesser,
+        ntry=2,
+    )
+    runner = Runner(
+        fitter=fitter,
+        guesser=guesser,
+        ntry=2,
+    )
+
+    mcal_kws = {'psf': 'gauss'}
+    if use_bootstrapper:
+        boot = MetacalBootstrapper(
+            runner=runner, psf_runner=psf_runner,
+            rng=rng,
+            **mcal_kws,
+        )
+        boot.go(obs)
+        resdict = boot.get_result()
+    else:
+        resdict = metacal_bootstrap(
+            obs=obs, runner=runner, psf_runner=psf_runner,
+            rng=rng,
+            **mcal_kws,
+        )
+
+    for key in ['noshear', '1p', '1m', '2p', '2m']:
+        assert key in resdict
+        assert resdict[key]['flags'] == 0
+
+
+@pytest.mark.parametrize('noise', [1.0e-8, 0.01])
+@pytest.mark.parametrize('use_bootstrapper', [False, True])
+def test_metacal_bootstrap_gaussmom_smoke(noise, use_bootstrapper):
+    """
+    Smoke test a Runner running the LM fitter
+    """
+
+    rng = np.random.RandomState(2830)
+
+    data = get_model_obs(
+        rng=rng,
+        model='gauss',
+        noise=noise,
+    )
+    obs = data['obs']
+
+    fwhm = 1.2
+    fitter = GaussMom(fwhm=fwhm)
+    psf_fitter = GaussMom(fwhm=fwhm)
+
+    psf_runner = PSFRunner(fitter=psf_fitter)
+    runner = Runner(fitter=fitter)
+
+    mcal_kws = {'psf': 'gauss'}
+    if use_bootstrapper:
+        boot = MetacalBootstrapper(
+            runner=runner, psf_runner=psf_runner,
+            rng=rng,
+            **mcal_kws,
+        )
+        boot.go(obs)
+        resdict = boot.get_result()
+    else:
+        resdict = metacal_bootstrap(
+            obs=obs, runner=runner, psf_runner=psf_runner,
+            rng=rng,
+            **mcal_kws,
+        )
+
+    for key in ['noshear', '1p', '1m', '2p', '2m']:
+        assert key in resdict
+        assert resdict[key]['flags'] == 0
```

Comment:
We should add something more quantitative here I think.

Reply:
what do you have in mind?

Reply:
I am not sure. Maybe we could test that the images and noise images that come out are different than the input and from each other? That is not a wonderful test, but we want to make sure the output data isn't obviously wrong.

Reply:
note the output observation does not have a noise image attached

Reply:
Ah right. We can still test the output image itself.

Reply:
I added image tests as well as a test of the recovered R value

Reply:
This is great!